apollo-gradle-plugin-kotlin/src/main/kotlin/com/apollographql/apollo/gradle/kotlin/ApolloPlugin.kt

```
@@ -0,0 +1,243 @@
+package com.apollographql.apollo.gradle.kotlin
+
+import com.android.build.gradle.AppExtension
+import com.android.build.gradle.LibraryExtension
+import com.android.build.gradle.api.BaseVariant
+import com.android.build.gradle.internal.tasks.factory.dependsOn
+import com.apollographql.apollo.compiler.GraphQLCompiler
+import org.gradle.api.Action
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.internal.HasConvention
+import org.gradle.api.internal.file.FileResolver
+import java.io.File
+import org.gradle.api.plugins.JavaPluginConvention
+import org.gradle.api.tasks.TaskProvider
+import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet
+import javax.inject.Inject
+import kotlin.IllegalArgumentException
+
+
+class ApolloPlugin : Plugin<Project> {
+  companion object {
+    const val TASK_GROUP = "apollo"
+  }
+
+  lateinit var apolloExtension: ApolloExtension
+
+  override fun apply(project: Project) {
+    apolloExtension = project.extensions.create("apollo", ApolloExtension::class.java)
+
+    // the extension block has not been evaluated yet, register a callback once the project has been evaluated
+    project.afterEvaluate {
+      afterEvaluate(it)
+    }
+  }
+
+  private fun afterEvaluate(project: Project) {
+    val generateClassesTask = project.tasks.register("generateApolloClasses") {
+      it.group = TASK_GROUP
+    }
+
+    val androidExtension = project.extensions.findByName("android")
+
+    apolloExtension.validate()
+
+    when {
```

nit: should it be
```
when(androidExtension) {
  is LibraryExtension ->
  is AppExtension -> 
  else ->
```

---

androidExtension could also be `InstantAppExtension` or something else. I'm not sure what to do in that case. I guess we can throw an exception

---

Hm, interesting, I guess for now we can say it's unsupported and revisit later.